### PR TITLE
Enhance README with examples, add Typescript definitions

### PR DESCRIPTION
**Pull request by orefalo**
Nice work, was just thinking about writing my own implementation. Thank you

just added typescript definitions.

Please publish on npm at your convenience.

### CODE REVIEW

**Comment by novemberborn** [comment]
@orefalo thanks for your PR! I'm hoping to have some time over the weekend to land this.

**Comment by orefalo** [comment]
You are welcome. plz also publish to npm - need the dependency badly.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/// <reference types="node" />
+export declare class KSUID {
+    constructor(buffer: any);
+    readonly date: Date;
+    readonly timestamp: number;
+    readonly payload: Buffer;
+    readonly string: string;
+    compare(other: KSUID): any;
```
Comment (novemberborn) [inline index.d.ts]: Can this be `-1`, `1` or `0`? Or alternatively `number`?

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    equals(other: KSUID): boolean;
+    toString(): string;
+    static random(): Promise<KSUID>;
+    static randomSync(): KSUID;
+    static fromParts(timeInMs: any, payload: any): KSUID;
```
Comment (novemberborn) [inline index.d.ts]: `any` should be `number` and `Buffer` respectively I think.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+/// <reference types="node" />
+export declare class KSUID {
+    constructor(buffer: any);
```
Comment (novemberborn) [inline index.d.ts]: `Buffer` rather than `any`?

**Comment by orefalo** [comment]
You are right, actually, I rewrote your library in pure typescript (check the typescript branch) to generate the d.ts.

Will have a look

**Comment by orefalo** [comment]
```
/// <reference types="node" />
export declare class KSUID {
    constructor(buffer: Buffer);
    readonly date: Date;
    readonly timestamp: number;
    readonly payload: Buffer;
    readonly string: string;
    compare(other: KSUID): number;
    equals(other: KSUID): boolean;
    toString(): string;
    static random(): Promise<KSUID>;
    static randomSync(): KSUID;
    static fromParts(timeInMs: number, payload: Buffer): KSUID;
    static isValid(buffer: Buffer): boolean;
    static parse(str: string): KSUID;
}
```

**Comment by orefalo** [comment]
Hi, sorry to be pushy, any updates on this pull req?

**Comment by novemberborn** [comment]
@orefalo I've landed the TypeScript changes in 640fbb6036f8fd821061de4dfbe88254ac9a3fb1. It's on npm as `0.3.0`.

**Comment by plfx** [comment]
👍  For the readme please.

**Comment by JoshFerge** [comment]
the readme is v. helpful, anything stopping it from being merged?

**Comment by novemberborn** [comment]
@JoshFerge it's helpful but it should be more complete. The API is much larger than this PR lets on. Would you be interested in doing a new PR for that?

**Comment by plfx** [comment]
Why not merge this PR to immediately improve the situation **and** take up separate work to do something more expansive.

**Comment by orefalo** [comment]
b/c it give credits to others...

**Comment by novemberborn** [comment]
@orefalo come on, that isn't fair.

This PR as it stands needs more fixing up, e.g. the examples use `require('./index')` rather than `require('ksuid')`. I started doing that, and then I realized that even with those fixes it misses so much of the API that having it is misleading as for what this library does. Hence the request to @JoshFerge to give it a go.

The time I have for OSS is limited and this project isn't my top priority. I'm very happy to take contributions and give appropriate credit.

**Comment by orefalo** [comment]
We align! It was up to date in June last year. ;-)
Relax, I am not trying to start a fight. Just pointing so frustration back in the time.
Topic closed. Great implementation thank you for your work.

**Pull request closed without merging**